File: MAY 24/18 May/Equalize the Towers gfg.cpp
class Solution{
    private:
    long long calculateCost(int n, int height, int cost[], int h[])
    {
        long long totalCost = 0ll;
        for(int i=0; i<n; i++)
            totalCost += 1ll*(abs(h[i]-height)*cost[i]);
        return totalCost;
    }

	public:
	long long int Bsearch(int n, int h[], int cost[]) 
	{ 
		// Your code goes here
		int low = *min_element(h, h+n), high=*max_element(h, h+n);
		for(int i=0; i<n; i++)
            high = max(high, h[i]);
		long long minCost = (1ll<<62), ans, eqHeight;
		long long lowCost, upCost;
		
		while(low<=high)
		{
	        eqHeight = (low+high)/2;
		    
		    minCost = calculateCost(n, eqHeight, cost, h);
		    upCost = calculateCost(n, eqHeight+1, cost, h);
		    lowCost = calculateCost(n, eqHeight-1, cost, h);
		  //  cout<<eqHeight<<" : "<<minCost<<", "<<lowCost<<", "<<upCost<<endl;
		    if(minCost<upCost && minCost<lowCost)
		    {
		        return minCost;
		    }
		    else if(lowCost<minCost)
		        high = eqHeight-1;
		    else
		        low = eqHeight+1;
		    ans = min(minCost, ans);
		}
		return ans;
		
	}
		 

};

File: JUNE 2025/16 June/Equalize the Towers gfg.java
class Solution {
    public long findCost(int[] heights, int[] cost, int height)
    {
        long total = 0, N = heights.length;
        for(int i=0; i<N; i++)
        {
            total += 1l*Math.abs(heights[i]-height)*cost[i];
        }
        return total;
    }
    public int minCost(int[] heights, int[] cost) {
        // code here
        int low = Arrays.stream(heights).min().getAsInt();
        int high = Arrays.stream(heights).max().getAsInt();
        
        int ans = Integer.MAX_VALUE, equalHeight;
        long currentCost, upper, lower;
        while(low<=high)
        {
            equalHeight = (low+high)/2;
            
            currentCost = findCost(heights, cost, equalHeight);
            lower = findCost(heights, cost, equalHeight-1);
            upper = findCost(heights, cost, equalHeight+1);
            
            ans = (int)Math.min(ans, currentCost);
            
            if(currentCost>lower)
                high = equalHeight-1;
            else if(currentCost>upper)
                low = equalHeight+1;
            else
                break;
        }
        return ans;
    }
}
